KEP-5547: Integrate Workload APIs with Job controller (#5871)
kubernetes/enhancements: keps/sig-apps/5547-integrate-workload-with-job/README.md

### Non-Goals

- Supporting dynamic changes to `minCount` or gang membership at runtime
- Complex workload structures with multiple nested PodGroups are not supported in alpha.

## Proposal
> This KEP depends on:
> - [KEP-4671: Gang Scheduling using Workload Object](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/4671-gang-scheduling)
> - [KEP-5832: Decouple PodGroup API from Workload API](https://github.com/kubernetes/enhancements/pull/5833)

The Job controller will be extended to create `Workload` and `PodGroup` objects as part of its pod management lifecycle. 
This integration ensures that pods belonging to a Job are scheduled according to the appropriate scheduling policy (gang or basic) before they are created. 

For the alpha release, this feature is optimized for static batch workloads with a flat API structure where 
`minCount` is immutable. The key design principles are:
- One `Job` creates one `Workload` with one `PodGroup` representing a single homogeneous group of pods.
- The automatic policy selection is based on `Job` Type
  - Jobs with `parallelism > 1` use gang scheduling policy where `minCount` equals the Job's parallelism.

>   - Jobs without indexed completion mode or `completions = 1`, use basic scheduling policy (pod-by-pod scheduling - `minCount`).

Reviewer: minCount?

  template:
    spec:
      containers:
      - name: trainer
        image: training-image:latest
        resources:
          limits:
            nvidia.com/gpu: 1
---
apiVersion: scheduling.k8s.io/v1alpha1
kind: PodGroup
metadata:
  name: pg-w-distributed-training
  namespace: training
  ownerReferences:
  - apiVersion: batch/v1
    kind: Job
    name: distributed-training
    uid: <job-uid>
    controller: true
spec:
  workloadRef:
    name: w-distributed-training
  podGroupTemplate:
    name: main
    schedulingPolicy:
      gang:
        minCount: 8  # Equal to Job.spec.parallelism
---
apiVersion: scheduling.k8s.io/v1alpha1
kind: Workload
metadata:
  name: w-distributed-training
  namespace: training
  ownerReferences:
  - apiVersion: batch/v1
    kind: Job
    name: distributed-training
    uid: <job-uid>
    controller: true
spec:
  podGroupTemplate:
  - name: pg-w-distributed-training
    policy:
      gang:
        minCount: 8  # Equal to Job.spec.parallelism
```

Then, the Job Controller will create the corresponding pods and set the `workloadRef` field:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: <job-name>-<random-suffix>
  namespace: training

>   ownerReferences:

Author: IMHO, we should add OwnerReference to PodGroup too.

>   ownerReferences:

Reviewer 2: Right - that's what I have in my mind.

The arrows in the diagram are misleading to me (e.g. it's not Job->Pod, but rather Pod->Job, same as Pod->PodGroup). Same for Job->Workload and Job->PodGroup.

But I agree with that - so let's just update the KEP to reflect all the ownerRefs in the examples.

- Feature gate `EnableWorkloadWithJob` is enabled by default
- Address feedback from alpha
- E2e tests covering gang scheduling scenarios
- Metrics for monitoring Workload/PodGroup creation and scheduling outcomes
- Performance testing to validate no significant impact on Job creation latency

#### GA

TBD after beta release

#### Deprecation

N/A for alpha release

### Upgrade / Downgrade Strategy

- **Upgrade:**
  1. Upgrade kube-apiserver
  2. Enable feature gate on kube-controller-manager
  3. New Jobs automatically get Workload/PodGroup objects
  4. Existing Jobs continue to work (no Workload created for them)

- **Downgrade:**
  1. Disable feature gate on kube-controller-manager
  2. New Jobs no longer get Workload/PodGroup objects
  3. Existing `Workload` and `PodGroup` objects remain
  4. Jobs with `workloadRef` on pods continue to run (field ignored)

- **Migration for Existing Jobs:**
  - Existing Jobs before upgrade do not automatically get Workload objects
  - To add gang scheduling to existing Jobs, delete and recreate them

### Version Skew Strategy

- kube-apiserver must be upgraded first to serve Workload API
- kube-controller-manager can be upgraded last

> If kube-controller-manager creates `Workload` but scheduler doesn't understand it:

Reviewer 2: Ahh - actually, those are different components, so it is not that simple. 

So we should talk about kube-scheduler a bit more here, I guess.

  - Pods will have `workloadRef` but scheduler ignores it
  - Standard pod-by-pod scheduling occurs with no gang scheduling benefit

If scheduler supports gang scheduling but controller doesn't create `Workload` objects, pods without `workloadRef` are scheduled normally with no gang scheduling benefit

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `EnableWorkloadWithJob`
  - Components depending on the feature gate:

>     - kube-apiserver

Reviewer 3: I think we would need kube-apiserver for defaulting or validation logic.

## Summary

This KEP introduces native integration between the Job controller and the gang scheduling[^1] APIs ([Workload](https://github.com/kubernetes/kubernetes/blob/release-1.35/pkg/apis/scheduling/types.go#L97) and [PodGroup](https://github.com/kubernetes/enhancements/pull/5833)).

The Job controller will automatically create `Workload` and `PodGroup` objects before creating pods for parallel Jobs, enabling native gang scheduling support in Kubernetes. 

## Motivation

The Kubernetes Job Controller currently creates pods independently without workload-aware scheduling constraints. This creates challenges for parallel applications (i.e., AI/ML training workloads, MPI jobs) that require all pods to be scheduled and run together or none(gang scheduling[^1]). Since there is now a native mechanism to express gang scheduling requirements now via `Workload` and `PodGroup` APIs, this KEP brings gang scheduling feature to Job Controller by integrating these APIs directly into the Job controller lifecycle.

### Goals

- Job controller automatically creates `Workload` and `PodGroup` objects for Jobs that require gang scheduling.

> - Job with `parallelism > 1` will use `GangSchedulingPolicy` with `minCount = parallelism`

Author: > Maybe ownerReferences check would be sufficient here.

This is what's stated in the KEP. You can find it in the notes/constraints section in addition to the unit tests and integration tests.

> - Job with `parallelism > 1` will use `GangSchedulingPolicy` with `minCount = parallelism`

Reviewer 4: Non-indexed jobs can make a workload with Basic policy.

> - Job with `parallelism > 1` will use `GangSchedulingPolicy` with `minCount = parallelism`

Reviewer 2: > Honestly, I'm inclined to start with stricter rules for creating the gang, and we can expand as we go, and we see it makes sense.

+1 to it
The goal is to prove the integration in Alpha, not to support every potential usecase for gang-scheduling. We should focus on not breaking anyone in this phase.

So the question is - what are the exact rules to use for gang-scheduling policy. Are we suggesting:
- parallelism > 1
- parallelism = completion
- Indexed only

[For everything else we can still create Workload, but use the Basic scheduling policy]

- Jobs that don't qualify for gang scheduling will use `BasicSchedulingPolicy`
- Ensure proper ordering of Workload/PodGroup creation before pods creation
- Existing Jobs without gang scheduling continue to work normally

### Non-Goals

- Supporting dynamic changes to `minCount` or gang membership at runtime
- Complex workload structures with multiple nested PodGroups are not supported in alpha.

> 

Reviewer 5: I believe that will depend on what we decide in the goals section.

## Proposal
> This KEP depends on:
> - [KEP-4671: Gang Scheduling using Workload Object](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/4671-gang-scheduling)
> - [KEP-5832: Decouple PodGroup API from Workload API](https://github.com/kubernetes/enhancements/pull/5833)

The Job controller will be extended to create `Workload` and `PodGroup` objects as part of its pod management lifecycle. 
This integration ensures that pods belonging to a Job are scheduled according to the appropriate scheduling policy (gang or basic) before they are created. 

For the alpha release, this feature is optimized for static batch workloads with a flat API structure where 
`minCount` is immutable. The key design principles are:
- One `Job` creates one `Workload` with one `PodGroup` representing a single homogeneous group of pods.
- The automatic policy selection is based on `Job` Type
  - Jobs with `parallelism > 1` use gang scheduling policy where `minCount` equals the Job's parallelism.
  - Jobs without indexed completion mode or `completions = 1`, use basic scheduling policy (pod-by-pod scheduling - `minCount`).

> - Elastic Jobs (changing parallelism at runtime) are not supported when gang scheduling is active.

Reviewer 3: I would prefer to forbid modifying parallelism in this case but do it in a way that doesn't break existing users of this feature.

> - Elastic Jobs (changing parallelism at runtime) are not supported when gang scheduling is active.

Author: Updated the proposal and added elastic job to the risks, constrains, and to beta criteria. PTAL

- Jobs created by higher-level controllers (i.e., JobSet) are skipped, the parent controller manages Workload/PodGroup lifecycle

> An example of the Job and the corresponding Workload/PodGroup creation:

Reviewer 2: I don't think naming conventions are enough. There are limitations for the length of names, so whatever conventions we will do, at some point we will hit it and will need to workaround it.
We should NOT rely purely on naming for discovery (e.g. check if I should create a Workload or one already exists).

So we have effectively two aspects here:

1) Discovery of `Workload` objects. Here we need to rely purely on references:

a) `Workload` has a `workload.spec.ControllerRef` field and job-controller (or in general controllers) should rely on that to find a `Workload` object for a given `Job`

b) If the `Workload` object was created by the job-controller (or in general controller) as opposed to end-user, it should also has an `ownerReference` with `Controller=true` set to the corresponding Job object.
Note that, I consciously wrote that it's only the case for objects created by the controller. If as a user, I brought my own custom `Workload`, then it should NOT have ownerReference set to the Job (ownerRef should not be adopted by the job-controller), becase we don't want to trigger deletion of the Workload object if the job gets deleted.
FWIW, this would also work, if the `Workload` was created by a higher-level controller.

2) Discovery of `PodGroup` objects. I think that we should try to mimic the above behavior

a) `PodGroup` will have a `PodGroupTemplateRef`, which is referencing the template. job-controller should rely on that to find a `PodGroup` object for a given `Job`.
Now, it works for single-PodGroup structures (our case), but we will need to address the aspect of how to decide which of the PodGroup from a given template a controller is looking for. But I don't think we need that for Alpha.

b) If the `PodGroup` object was created by the job-controller, as opposed to some other entity, it should have two ownerReferences set:
- to the Job (with Controller=true)
- to the Workload (without Controller=true)
In this case we want to ensure that if both Workload and Job gets deleted, the PodGroup will get deleted too.
If we're adopting PodGroup created by the user/higher-level controller, the no ownerReferences should be added by the controller.

3) Naming convention
Due to limitations related to naming, it will not be possible to rely on that for discovery. However, it will still be useful to have naming that allows for humans to create logical links between objects.
We already have some prior-art in this area, e.g. in Deployment:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/sync.go#L560-L568
and we should learn from it.
But OTOH, our case is also different due to not 1-1 relation as mentioned by Kevin in the first comment.

My personal proposal would be go with something simple first:

a) For Workload:
- the name of workload is: `<(truncated-if-needed)job-name>-<hash>`
- the hash allows for collisions handling (maybe we can simply rely on "generatedName" even?)
- for distinguishing between different types, we rely on other metadata (OwnerReferences has type field, alternatively maybe we can have a dedicated annotation/label for it?)
Given that we will not rely on discovery for naming, we can probably change the pattern later.

b) For PodGroup:
- the name of podgroup is: `<(truncated-if-needed)workload-name>-<(truncated-if-needed)podgroup-name>-<hash>`
- the hash allows for multiple groups within a Workload and PodGroupTmemplate
Again, given we don't rely on naming for discovery, we probably can change the pattern later.

@soltysh  - would that address your concerns?

> An example of the Job and the corresponding Workload/PodGroup creation:

Reviewer 5: > If as a user, I brought my own custom Workload, then it should NOT have ownerReference set to the Job (ownerRef should not be adopted by the job-controller), becase we don't want to trigger deletion of the Workload object if the job gets deleted.

But this still assumes that some other `ownerRefs` with `controller=true` are set, so that we know that we should not adopt it, do I understand that correctly?

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: distributed-training
  namespace: training
spec:
  parallelism: 8
  completions: 8
  completionMode: Indexed
  template:
    spec:
      containers:
      - name: trainer
        image: training-image:latest
        resources:
          limits:
            nvidia.com/gpu: 1
---
apiVersion: scheduling.k8s.io/v1alpha1
kind: PodGroup
metadata:
  name: pg-w-distributed-training
  namespace: training
  ownerReferences:

>   - apiVersion: batch/v1

Author: I'd go with the Job. I would like to stick the `PodGroup` object with the controller through the lifecycle. wdyt?

    kind: Job
    name: distributed-training
    uid: <job-uid>
    controller: true
spec:
  workloadRef:

>     name: w-distributed-training

Reviewer 2: Let's not rely explicitly on naming - see my comment above.

  podGroupTemplate:
    name: main
    schedulingPolicy:
      gang:
        minCount: 8  # Equal to Job.spec.parallelism
---
apiVersion: scheduling.k8s.io/v1alpha1
kind: Workload
metadata:
  name: w-distributed-training
  namespace: training
  ownerReferences:
  - apiVersion: batch/v1
    kind: Job
    name: distributed-training
    uid: <job-uid>
    controller: true
spec:
  podGroupTemplate:
  - name: pg-w-distributed-training
    policy:
      gang:
        minCount: 8  # Equal to Job.spec.parallelism
```

Then, the Job Controller will create the corresponding pods and set the `workloadRef` field:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: <job-name>-<random-suffix>
  namespace: training
  ownerReferences:
  - apiVersion: batch/v1
    kind: Job
    name: distributed-training
spec:
  workloadRef:
    name: w-distributed-training

>     podGroupName: pg-w-distributed-training

Reviewer 6: ```suggestion
  workloadRef:
    name: distributed-training
    podGroupName: distributed-training
```

  containers:
  - name: ...
```

### User Stories (Optional)

#### ML Training Job with Gang Scheduling

As a machine learning engineer, I want to run a distributed training job with 8 workers that must all be scheduled together. 
If only 7 workers can be scheduled, I don't want any pods to start because partial training would waste resources.

#### Standard Batch Job with Workload Tracking

As a data engineer, I want to run a batch processing job that processes files sequentially without gang scheduling requirements.

### Notes/Constraints/Caveats (Optional)

#### Alpha Constraints

- The alpha release targets simple, static batch workloads where the workload requirements are known at creation time.
- Each Job maps to one `PodGroup`. All pods in the Job are identical from a scheduling policy perspective.
- The `minCount` field in the Workload's `GangSchedulingPolicy` mirrors the Job's parallelism.
- There is no mechanism to opt-out of `Workload`/`PodGroup` creation for indexed (parallel) jobs if feature gate is enabled.

> - When gang scheduling is active (parallel jobs), changes to `spec.parallelism` are blocked via admission validation because this would require changing `minCount`

Reviewer 6: As I mentioned above, I don't understand why do we need this limitation?

> - If a Job has `ownerReferences` indicating it is managed by another controller (i.e., JobSet), the Job controller 

Reviewer 6: ```suggestion
- If a Job has `ownerReferences` indicating it is managed by another controller (i.e., JobSet, TrainJob), the Job controller 
```

Reviewer 2: > But at the same time this means that a CronJob-owned Jobs will not be capable of using Workloads

I would say it's desired - let's not try to boil the ocean here.

will not create `Workload`/`PodGroup` objects.

### Risks and Mitigations

## Design Details

### Job Controller Changes

The Job controller reconciliation loop for processes each Job will be extended to ensure `Workload` and `PodGroup` objects exist before creating pods. The modified workflow proceeds as follows:

- Check if a `Workload` object already exists for this `Job`.
  - If not, determine the appropriate scheduling policy and create the `Workload` object with the determined policy.
  - If it already exists, verify the existing `Workload` matches the `Job` spec. If not, update the `Workload` object.
- Check if a `PodGroup` object already exists for this `Job`.
  - If not, create the `PodGroup` object referencing the `Workload`

>   - If it already exists, verify the existing `PodGroup` correctly references the `Workload`.

Author: Rewrote this section to address your feedback. PTAL.

- Execute existing pod management logic to create pods, include `workloadRef.podGroupName` in the pod spec to associate pods with the `PodGroup`.

The Job Controller will create `Workload` and `PodGroup` based on Job configuration:
 - `parallelism > 1`: `GangSchedulingPolicy` with `minCount=parallelism`
 - `parallelism = 1` (default): `BasicSchedulingPolicy`
 - `completions = 1, parallelism = 1`: `BasicSchedulingPolicy`
 - `completions > 1, parallelism = 1`: `BasicSchedulingPolicy`

The controller will require additional informers and listers for `Workload` and `PodGroup` objects. Both `Workload` and `PodGroup` are automatically garbage collected when the `Job` is deleted.

#### Object Creation Order

The Job controller must create objects in a strict order to ensures that the scheduler can properly validate pods 
against their scheduling policy before attempting to schedule them. The order is as follows:

> 1. `Workload` object

Reviewer 6: Is that really needed? I asked previously, it doesn't matter in which order objects will be created since kube-scheduler will wait for Workload and PodGroup objects if Pods have workloadRef

2. `PodGroup` object that will references the `Workload`

> 3. `Pod` objects which will reference `PodGroup` and `Workload`

Reviewer 3: I was under the impression that PodGroup names should be auto generated as each instance of PodGroup creation should be unique as it may have some life cycle implications (ResourceClaims are only allocated/deallocated upon removal of PodGroup).

So I think that:
1. Job creates Workload with the name of Job
2. Job creates a PodGroup with random name (or GenerateName idk)
3. Get name of PodGroup and create pods that link to the podGroup you created.

If pods get deleted I think we should create a new PodGroup (maybe?).

This was the major question i have around pod lifecycle and how controllers should manage the lifecycle of PodGroups.

> 3. `Pod` objects which will reference `PodGroup` and `Workload`

Reviewer 6: That's right, the same name won't work if we want to support multiple PodGroups for Job in the future.
Also, we have naming problem which is identified here: https://github.com/kubernetes/enhancements/pull/5871#discussion_r2774841523
E.g. different resources in the same namespaces have the same name (e.g. TrainJob / Job).

We might want to go ahead with `generateName` field, although I am not sure.

> 3. `Pod` objects which will reference `PodGroup` and `Workload`

Author: Added subsection for Naming conventions based on @wojtek-t feedback. PTAL.

### Admission Validation for Parallelism Changes

When gang scheduling is active, the Job controller relies on admission validation to block changes to `Job.spec.parallelism`. Since changing this field would require changing `minCount` in the `Workload` object, which is immutable.

> 

Author: Added deletion/GC subsection and updated the risks/constrains for the suspended Jobs. PTAL.

- `k8s.io/kubernetes/pkg/controller/job_controller`: `2026-01-29` - `89.1%`
- `k8s.io/kubernetes/pkg/apis/batch/validation`:
- `k8s.io/kubernetes/pkg/registry/batch/job`: 
- Add test that verifies 
  - SchedulingPolicy for various Job configurations
  - Workload and PodGroup creation
  - pod creation includes correct `workloadRef`
  - Parallelism change is blocked for gang-scheduled Jobs and allowed for basic-scheduled Jobs
  - Job deletion cascades to Workload and PodGroup deletion
  - Feature gate disabled: Jobs work without Workload/PodGroup creation
  - Jobs with ownerReferences (managed by higher-level controllers) do not create Workload/PodGroup

##### Integration tests

We will add the following integration tests to the Job controller `https://github.com/kubernetes/kubernetes/blob/v1.35.0/test/integration/job/job_test.go`:
- Gang and Basic Scheduling Lifecycle Test (create, update, delete Job, verify Workload and PodGroup creation, verify pods have workloadRef, verify Job deletion cascades to Workload and PodGroup deletion)
- Failure Recovery Test (create Job with Workload API unavailable, verify Job controller retries, verify Workload is eventually created)

> - Feature gate disable/enable (Jobs work without Workload/PodGroup creation (Jobs with ownerReferences managed by higher-level controllers do not create Workload/PodGroup))

Author: The 2 and 3 already stated in the Proposal section already. I'll add the suyspended jobs.

> - Feature gate disable/enable (Jobs work without Workload/PodGroup creation (Jobs with ownerReferences managed by higher-level controllers do not create Workload/PodGroup))

Author: Addressed. PTAL.

- Failure scenarios, i.e., insufficient resources for gang, partial failures

### Graduation Criteria

#### Alpha (v1.36)

- Feature is implemented behind feature gate `EnableWorkloadWithJob` (default: disabled)
- Job controller creates `Workload` and `PodGroup` objects for Jobs when feature gate is enabled
- Job controller sets `podGroupKind: PodGroup` in pod specs, opting into the explicit runtime PodGroup model
- Gang scheduling policy applied to indexed parallel Jobs (`parallelism > 1` with `Indexed` completion mode)
- Basic scheduling policy applied to all other Job types
- Jobs managed by higher-level controllers skip Workload/PodGroup creation
- Admission validation blocks `parallelism` changes for gang scheduling Jobs
- Unit tests for all new Job controller logic
- Integration tests for Workload/PodGroup creation flow
- Documentation for enabling and using the feature

#### Beta
- Feature gate `EnableWorkloadWithJob` is enabled by default
- Address feedback from alpha
- E2e tests covering gang scheduling scenarios
- Metrics for monitoring Workload/PodGroup creation and scheduling outcomes
- Performance testing to validate no significant impact on Job creation latency

> 

Reviewer 5: Add here explicitly that for beta we need to have `minCount` being mutable to support Elastic Indexed Jobs.

No.

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

- `kubectl get workloads -A` will show `Workload` objects created by the Job controller
- `kubectl get podgroups -A` will show `PodGroup` objects created by the Job controller

###### How can someone using this feature know that it is working for their instance?

- [x] Events
  - Event Reason: `WorkloadCreated` - Emitted when `Workload` object is created for a Job
  - Event Reason: `PodGroupCreated` - Emitted when `PodGroup` object is created for a Job
- [ ] API .status
  - Condition name: 
  - Other field: 
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

 - 99.9% of Workload/PodGroup creations succeed within 10 seconds

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [x] Metrics
  - Metric name:
  - [Optional] Aggregation method:
  - Components exposing the metric:
- [ ] Other (treat as last resort)
  - Details:

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

> 

Reviewer 5: >     * job_controller_workload_creations_total
>     * job_controller_podgroup_creations_total

They should match existing pattern, so it'll be more like `job_controller_job_workloads_creation_total` and `job_controller_job_podgroups_creation_total`

>     * job_controller_workload_creation_duration_seconds

I don't think we need this particular one. 

>     * job_controller_workload_skip_total

I believe we should be able to calculate that value based on no of existing jobs - no of created workloads. So I think we can skip that too.

> 

Author: Sure. Do you think we should add these two metrics and state they will be implemented for alpha?

> 

Reviewer 5: Let's hold off with adding them, and reconsider its value during implementation. Those are not required for alpha, so no need to rush that in. But definitely worth keeping in mind.

The Job controller will automatically create `Workload` and `PodGroup` objects before creating pods for parallel Jobs, enabling native gang scheduling support in Kubernetes. 

## Motivation

The Kubernetes Job Controller currently creates pods independently without workload-aware scheduling constraints. This creates challenges for parallel applications (i.e., AI/ML training workloads, MPI jobs) that require all pods to be scheduled and run together or none(gang scheduling[^1]). Since there is now a native mechanism to express gang scheduling requirements now via `Workload` and `PodGroup` APIs, this KEP brings gang scheduling feature to Job Controller by integrating these APIs directly into the Job controller lifecycle.

### Goals

- Job controller automatically creates `Workload` and `PodGroup` objects for Jobs that require gang scheduling
- Support opt-out mechanism to define when Job controller skips creating `Workload` and `PodGroup` objects
- Use`GangSchedulingPolicy` with `minCount = parallelism` for Jobs with `parallelism > 1`, `completionMode: Indexed`, and `parallelism = completions`.

> - Jobs that don't qualify for gang scheduling will use `BasicSchedulingPolicy` and still get `Workload` and `PodGroup` objects created.

Author: Creating `Workload` and `PodGroup` for non-gang (`BasicSchedulingPolicy`): 

- Gives the workload-aware workstream a single, consistent model
- Keeps a single code path in the Job controller and scheduler, every Job gets a Workload and PodGroup, and every pod gets a scheduling group reference. User won't get confused why this Job didn't get a Workload object?
- Any future scheduling (and potentially more like topology-aware, quota, etc...) features expressed at the Workload level, which can benefit gang and non-gang Jobs. If only gang Jobs had these objects, those features would either not apply to basic Jobs or would require a second code path. 

@wojtek-t  @mm4tt Please feel free to elaborate more.

btw, should we add this to the KEP?

- Jobs created by CronJob do not have `workloadRef` set, the Job controller creates one `Workload` and one `PodGroup` per Job for them.

An example of the Job and the corresponding `Workload`/`PodGroup` creation:

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: <job-name>
  namespace: training
spec:
  parallelism: 8
  completions: 8
  completionMode: Indexed
  template:
    spec:
      containers:
      - name: trainer
        image: training-image:latest
        resources:
          limits:
            nvidia.com/gpu: 1
---
apiVersion: scheduling.k8s.io/v1alpha1
kind: Workload
metadata:
  name: <job-name>-<hash>
  namespace: training
spec:
  controllerRef:
    apiVersion: batch/v1
    kind: Job
    name: <job-name>
    uid: <job-uid>
  podGroupTemplate:
  - name: pg-w-distributed-training
    policy:
      gang:
        minCount: 8  # Equal to Job.spec.parallelism
---
apiVersion: scheduling.k8s.io/v1alpha1
kind: PodGroup
metadata:
  name: <workload-name>-podgroup-name-<hash>
  namespace: training
  ownerReferences:
  - apiVersion: batch/v1
    kind: Job
    name: <job-name>
    uid: <job-uid>

>     controller: true

Reviewer 5: Based [on this comment](https://github.com/kubernetes/enhancements/pull/5871#discussion_r2782195420), you're missing here a reference to workload. Or we decided it's not needed?

spec:
  podGroupTemplateRef:
    workloadName: <workload-name>
    podGroupTemplateName: <podgroup-template-name>
  schedulingPolicy:
    gang:
      minCount: 8  # Equal to Job.spec.parallelism
```
Then, the Job Controller will create the corresponding pods and set the `workloadRef` field:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: <job-name>-<random-suffix>
  namespace: training
  ownerReferences:
  - apiVersion: batch/v1
    kind: Job
    name: <job-name>
    uid: <job-uid>
spec:

>   workloadRef:

Reviewer 2: Let's update it based on discussions on Workload/PodGRoup KEPS.

- When gang scheduling is active (`GangSchedulingPolicy`), changes to `spec.parallelism` (scaling up/down) are rejected via conditional validation (depends on feature enablement). This is because this would require changing `minCount` in the `Workload` object, which is immutable. This will disable [Elastic Indexed Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/job/#elastic-indexed-jobs).

### Risks and Mitigations

- `JobSet` or other higher-level controllers will create `Workload`/`PodGroup` objects for their `Workload`, the Job controller will duplicate or update the objects and create a conflict. We can mitigate this by ensuring the Job controller will not create `Workload`/`PodGroup` objects if the Job has `ownerReferences` indicating it is managed by higher-level controller.

- When the feature is enabled, Jobs that identify as gang scheduling cannot have `spec.parallelism` changed. That effectively disables [Elastic Indexed Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/job/#elastic-indexed-jobs) for those Jobs. This is accepted for alpha with clear documentation and a committed path to beta to not break Elastic Indexed Jobs.

> - Suspended Jobs and resource release rely only on GC, which does not address releasing resources (DRA) while a Job is suspended. This behavior is acceptable for alpha. Future work may require the controller to delete `PodGroup`/`Workload` on suspend and recreate on resume.

Reviewer 3: IMO this is not a "Future work" but a must for beta.

A suspended job should clean up workloads and podgroups. And when it is resumed those resources get created.

## Design Details

### Job Controller Changes

The Job controller reconciliation loop for processes each Job will be extended to ensure `Workload` and `PodGroup` objects exist before creating pods. 

#### Workload and PodGroup Discovery

Discovery of those objects is based on references (`workload.spec.controllerRef` and `podGroup.spec.podGroupTemplateRef`), not on ownership. `ownerReference` is used only for controller-created objects so that they are garbage-collected when the Job is deleted. Workloads which are created by user or higher-level controller are not given ownerReferences to the Job, so they are not deleted when the Job is deleted.

A `Workload` is considered the Workload for this Job object if:
- The `Workload` is in the Job’s namespace
- It has `workload.spec.controllerRef` field that is associated with this Job

Similarly, a `PodGroup` is considered the `PodGroup` for this Job if:
- The `PodGroup` is in the Job’s namespace

> - It has a controller `ownerReference` pointing to this Job

Reviewer 2: Why this one?

As already stated above, ownerRefs should be used for lifecycle management and/or ownership, not discovery.

Author: Removed.

The controller will require additional informers and listers for `Workload` and `PodGroup` objects. Both `Workload` and `PodGroup` are automatically garbage collected when the `Job` is deleted.

If the `Workload` was created by someone else and the scheduling policy was set to `GangSchedulingPolicy` while it shouldn't (and vice versa) or other cases that controller doesn't expect, the Job controller will treat that as an opt-out, the `Workload` and `PodGroup` objects will be ignored and the Job will continue to work with normal workflow. A condition or event should be triggered to inform the user. 

#### Object Creation Order

The Job controller creates objects in the following order so that references point to existing objects and to satisfy any API validation that `Workload` exists before `PodGroup` is created. The order is as follows:
1. `Workload` object
2. `PodGroup` object that will references the `Workload`
3. `Pod` objects which will reference `PodGroup` and `Workload`

The kube-scheduler waits for `PodGroup` when Pods have `workloadRef`, so scheduling does not depend on this order, the order is for consistency and API validity.

### Validation for Parallelism Changes

> The Job API Validation rejects updates that change `spec.parallelism` when the feature gate is enabled and the Job uses gang scheduling. Since changing this field would require changing `minCount` in the `Workload` object, which is immutable.

Reviewer 5: > the Job uses gang scheduling

How are you going to ensure this part of the contract in validation? I assume the only way is to mimic the criteria from earlier:

> parallelism > 1` and `completionMode: Indexed` and `parallelism = completions`

or you have other idea?

### Naming Conventions

We will not use naming for discovery due to limitations related to naming. Naming is for human readability and logical linking between Job, `Workload`, and `PodGroup`. Because discovery does not depend on it, the naming pattern can be changed in later releases if needed. 

Following prior-art in [Deployment](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/sync.go#L560-L568), the naming convention can be as follows:

**1. Workload**
  - Pattern: `<(truncated-if-needed)job-name>-<hash>`
  - Truncation of the Job name is applied when necessary to respect object name length limits.
  - The hash is used for collision avoidance (implementation may use a generated suffix or a hash of relevant identity).
  - Object type (`Workload` vs `PodGroup`) is identified by other metadata (`ownerReferences[].kind`), not by the name pattern.

**2. PodGroup**

>   - Pattern: `<(truncated-if-needed)workload-name>-<(truncated-if-needed)podgroup-name>-<hash>`

Reviewer 5: I'm not complaining, especially that I don't have a better idea :sweat_smile: I agree that most names are short, but then again we've had issues with longer names in the past, so at least adding a test to ensure we're not making the names unreadable and most importantly that this doesn't cause unexpected problems is desirable.

- Elastic Indexed Jobs must be supported (beta blocker)
- The controller needs a mechanism to delete `PodGroup`/`Workload` in the case of suspended Jobs and recreate on resume.
- Feature gate `EnableWorkloadWithJob` is enabled by default
- Address feedback from alpha
- E2e tests covering gang scheduling scenarios
- Metrics for monitoring `Workload`/`PodGroup` creation and scheduling outcomes
- Performance testing to validate no significant impact on Job creation latency

#### GA

TBD after beta release

#### Deprecation

N/A for alpha release

### Upgrade / Downgrade Strategy

- **Upgrade:**
  1. Upgrade kube-apiserver
  2. Enable feature gate and upgrade kube-controller-manager
  3. New Jobs automatically get `Workload`/`PodGroup` objects
  4. Existing Jobs continue to work (no `Workload` created for them)

- **Downgrade:**
  1. Disable feature gate and downgrade kube-controller-manager
  2. New Jobs no longer get `Workload`/`PodGroup` objects
  3. Existing `Workload` and `PodGroup` objects remain

>   4. Jobs with `workloadRef` on pods continue to run (field ignored)

Reviewer 2: I'm assuming that new pods created for those jobs will not have the podGroupRef set, right?

- **Migration for Existing Jobs:**
  - Existing Jobs before upgrade do not automatically get `Workload` objects

>   - To add gang scheduling to existing Jobs, delete and recreate them

Author: Correct. Added.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `EnableWorkloadWithJob`
  - Components depending on the feature gate:
    - kube-controller-manager
- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node?

###### Does enabling the feature change any default behavior?
Yes. When the feature gate is enabled:

1. The Job controller creates `Workload` and `PodGroup` objects for all Indexed parallel Jobs (`parallelism > 1`, `completions = parallelism`, `completionMode: Indexed`) before creating pods.
2. Jobs (`parallelism > 1` with `Indexed` completion mode) use gang scheduling, meaning all pods must be scheduled together or none are scheduled.
3. Updates to `Job.spec.parallelism` are rejected for Jobs using gang scheduling.

> 4. Pod creation is delayed until the `PodGroup` object is acknowledged by the scheduler.

Author: Rewarded.PTAL

- If the API server doesn't support `Workload` and `PodGroup` APIs, the Job controller will fail to create Jobs (error creating `Workload`). Jobs will be requeued until the API server is upgraded.
- If the scheduler doesn't have gang scheduling feature enabled, pods will remain pending indefinitely.
- Already running Jobs are not affected by enabling the feature. Pods that are already scheduled and running continue to run. New Jobs or Jobs being reconciled will be affected.
- For the rollback, disabling the feature gate allows Jobs to work without `Workload` and `PodGroup` creation. `Workload` and `PodGroup` objects don't cause issues; they're ignored when the feature is disabled

###### What specific metrics should inform a rollback?

The following metrics should be monitored:

- `job_sync_duration_seconds`: If Job sync duration increases significantly, it may indicate issues with `Workload`/`PodGroup` creation
- `job_pods_creation_total`: If pod creation rate drops, gang scheduling may be blocking pod creation

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

> Yes. This will be tested as part of alpha release validation.

Author: Updated.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No.

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

- `kubectl get workloads -A` will show `Workload` objects created by the Job controller
- `kubectl get podgroups -A` will show `PodGroup` objects created by the Job controller

###### How can someone using this feature know that it is working for their instance?

- [x] Events
  - Event Reason: `WorkloadCreated` - Emitted when `Workload` object is created for a Job
  - Event Reason: `PodGroupCreated` - Emitted when `PodGroup` object is created for a Job
- [ ] API .status
  - Condition name: 
  - Other field: 
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

>  - 99.9% of `Workload`/`PodGroup` creations succeed within 10 seconds

Reviewer 2: Let's remove it for now and discuss it later. Same for metrics below.

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [x] Metrics
  - Metric name:
  - [Optional] Aggregation method:
  - Components exposing the metric:
- [ ] Other (treat as last resort)
  - Details:

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

- `scheduling.k8s.io/v1alpha1` for `Workload` and `PodGroup` APIs
- `kube-scheduler` with gang scheduling feature enabled

### Scalability

###### Will enabling / using this feature result in any new API calls?

Yes. For each Job, the following additional API calls are made:
- `GET Workload` - 1 per Job sync
- `CREATE Workload` - 1 per Job creation
- `GET PodGroup` - 1 per Job sync
- `CREATE PodGroup` - 1 per Job creation

> - `PATCH PodGroup.status` - 1 per scheduling decision

Author: updated. PTAL.

The controller will require additional informers and listers for `Workload` and `PodGroup` objects. Both `Workload` and `PodGroup` are automatically garbage if they were created by the job-controller and the corresponding Job is deleted.

If the `Workload` was created by another actor (i.e. JobSet, User creates a `Workload` with `BasicSchedulingPolicy` 
to opt-out of gang scheduling), the Job controller respects and uses it and its associated PodGroups. It does 
not treat that as an opt-out. The Job controller uses the discovered `Workload` and PodGroup when creating pods.
The Job controller falls back (ignore the discovered `Workload` and `PodGroup`) when the discovered `Workload` 
has an unsupported structure (for alpha, when the number of `PodGroupTemplates != 1`). In that case, a condition 
or event should be triggered to inform the user.

#### Object Creation Order

The Job controller creates objects in the following order so that references point to existing objects and to satisfy any API validation that `Workload` exists before `PodGroup` is created. The order is as follows:

> 1. `Workload` object

Reviewer 5: ```suggestion
1. `Workload` object which will reference the `Job`
```

> 2. `PodGroup` object that will references the `Workload`

Reviewer 5: ```suggestion
2. `PodGroup` object which will reference the `Workload` and the `Job`
```

- kube-controller-manager can be upgraded before or after the scheduler
- kube-scheduler should be upgraded before or together with kube-controller-manager. Gang semantics will not apply until the scheduler is upgraded.

There are different Skew scenarios involving the kube-scheduler:
- kube-controller-manager new, kube-scheduler old: The controller creates `Workload` and `PodGroup` and sets `schedulingGroup` on pods but the scheduler does not understand these objects and ignores `schedulingGroup`. In this case, pods are scheduled normally (pod-by-pod) with no gang scheduling benefit until the scheduler is upgraded.
- kube-controller-manager old, kube-scheduler new: The controller does not create `Workload`/`PodGroup` and does not set `schedulingGroup` on pods. The scheduler has no workload information for those Jobs. Pods are scheduled normally with no gang scheduling benefit.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `EnableWorkloadWithJob`
  - Components depending on the feature gate:

>     - kube-controller-manager

Reviewer 2: And kube-apiserver - that runs the admission logic.

Might be worth mentioning that given we need to support elastic jobs for Beta, kube-apiserver will eventually be removed from that list.
